Enable unprivileged image builds

DaanDeMeyer opened the pull request:
To enable this, when doing a build, we unshare a user namespace
with it's own private set of uids/gids obtained using newuidmap
and newgidmap. We also map the current user to the last UID/GID
in the UID/GID range from /etc/subuid and /etc/subgid. Together
with unsharing the mount namespace, this allows us to do
unprivileged bind and overlay mounts.

Next, we replace all usages of systemd-nspawn during the image build
with bubblewrap. systemd-nspawn cannot run as an unprivileged user
yet so we use bubblewrap which can. bubblewrap can also be used to
setup a chroot environment with API VFS filesystems so we make use
of that to setup chroot environments and remove all our homegrown
logic for it. This allows us to significantly reduce the amount of
mounts we do in mkosi itself.

To further reduce the amount of mounts, we modify the invocations
of all package managers to specify the cache directory via the
relevant option instead of mounting the cache directory into the
chroot. For apt, to accomplish this, we switch from using
DPkg::Chroot-Directory to setting the "--root" option for each
invocation of dpkg so that dpkg can access files outside of the
chroot.

Finally, we remove some options which become obsolete with this
commit, --idmap, --chown and --nspawn-keep-unit.

We also remove --source-file-transfer, --source-file-transfer-final
and the corresponding symlink options. Instead, we default to mounting
source files into the build tree. In the future, we'll add virtiofsd
support to allow accessing source files in qemu VMs.

We also move stuff around and create a few new files to store
helpers to avoid circular imports. There's also a little bit of
refactoring and cleanup all around.

DaanDeMeyer commented:
@257 Do you have a clue about the gentoo failures?

257 commented:
maybe this:
https://github.com/systemd/mkosi/pull/1313/files#diff-e89c10e3e1a25022797c92a8ba44fda97927bbb6a59220a48084d3bf70378de8L174

`CAP_SYS_ADMIN,CAP_MKNOD`

behrmann commented on mkosi/__init__.py:
```diff
@@ -4354,6 +4276,40 @@ def needs_build(config: Union[argparse.Namespace, MkosiConfig]) -> bool:
+    child = os.fork()
+    if child == 0:
+        event.wait()
+        run(["newuidmap", str(pid), "0", str(os.getuid()), "1", "1", str(subuid), str(int(subuidcount) - 1)])
+        run(["newgidmap", str(pid), "0", str(os.getuid()), "1", "1", str(subgid), str(int(subgidcount) - 1)])
```
```suggestion
        run(["newuidmap", f"{pid}, "0", f"{os.getuid()}", "1", "1", f"{subuid}", f"{subuidcount - 1}"])
        run(["newgidmap", f"{pid}", "0", f"{os.getuid()}", "1", "1", f"{subgid}", f"{subgidcount - 1}"])
```
Feels like less obstrusive than the `str` calls, also `subuidcount` and and `subgidcount` are already ints coming out of `read_subrange`.

Also this code might make more sense in `run.py`.

behrmann replied:
Not addressed.

DaanDeMeyer replied:
To be fair, I disagree here, I think  using `str()` is much nicer than using f strings here

DaanDeMeyer replied:
Addressed the other thing

behrmann replied:
> To be fair, I disagree here, I think using `str()` is much nicer than using f strings here

Totally fair, this actually was more about the `int` calls.

behrmann reviewed (commented):
Looks really nice!

257 commented:
@DaanDeMeyer did that help?

DaanDeMeyer commented:
@257 I've been busy fixing some other issues, haven't had a time to check whether it works yet. I don't think it will though, since we're running as root in the user namespace, so we should already have all capabilities.

DaanDeMeyer commented:
So --format directory images have issues booting because systemd-nspawn does not map the directory owner to root in the container. However, even if it did, there's still the issue that there might be files owned by other users in the directory tree, which are mapped to UIDs outside the valid UID range due to us using user namespacing with `newuidmap` and `newgidmap`.  The only way we'll get the directory tree booting properly would be to use the exact same idmapping used when building the image. I first figured we could force chown all deviating stuff to the current user and rely on tmpfiles to fix things on first boot but then we'll still end up with files owned by random UIDs on the host afterwards.

DaanDeMeyer commented:
@257 Can you rework the gentoo implementation to store logs and such in `state.workspace` instead of in the root filesystem? Doing that won't work anymore when running as an unprivileged user.

257 commented:
will have a look

DaanDeMeyer commented:
Reworked to use bubblewrap instead of unshare. unshare doesn't set up the API VFS properly and it turns out doing that ourselves isn't trivial. bubblewrap does what unshare does and has option that allow us to easily configure a API VFS for a chroot which simplifies the code a lot.

Also reworked run_build_script() to use run_workspace_command() which was long overdue. All package managers now use a new function run_with_apivfs() which sets up a chroot using bubblewrap and then invokes the passed command in it.

behrmann commented on mkosi/__init__.py:
```diff
@@ -3581,75 +3498,49 @@ def run_build_script(state: MkosiState) -> None:
         return
 
-    idmap_opt = ":rootidmap" if nspawn_id_map_supported() and state.config.idmap else ""
+    os.makedirs(install_dir(state), mode=0o755, exist_ok=True)
```
Any reason against
```suggestion
    install_dir(state).mkdir(mode=0o755, exist_ok=True)
```

behrmann commented on mkosi/distributions/__init__.py:
```diff
@@ -29,3 +32,8 @@ def remove_packages(cls, state: "MkosiState", remove: list[str]) -> None:
         raise NotImplementedError
+
+
+def cache_params(state: "MkosiState") -> list[PathString]:
```
This is quite specific to using bubblewrap, no? I guess that would be better places in `run.py`.

DaanDeMeyer commented:
Another rework, I've given up on trying to make all files on the host owned by the user running mkosi, it just doesn't work when using directory output. Instead, the current user is mapped to the same uid in the user namespace, and then we do the same as we do when running as root, we just chown the necessary stuff when needed. Also, to avoid chowning more than necessary, when possible we run processes as the current user where applicable.

With this rework, when using --format=directory, we end up with a directory tree on the host completely owned by the first UID in the private UID range from /etc/subuid. This directory tree is usable with `systemd-nspawn --private-users` so it can be booted in nspawn, unlike the previous approach where that wasn't possible.

behrmann commented on mkosi/__init__.py:
```diff
@@ -202,7 +179,8 @@
 
-            d = tempfile.TemporaryDirectory(dir=p, prefix=f"mkosi.{config.build_sources.name}.tmp")
+            d = tempfile.TemporaryDirectory(dir=p, prefix=f".mkosi.{config.build_sources.name}.tmp")
+        os.chmod(d.name, 0o755)
```
This should get a comment why the permisson needs to be like this.

behrmann commented on mkosi/__init__.py:
```diff
@@ -608,22 +562,8 @@ def configure_serial_terminal(state: MkosiState, cached: bool) -> None:
-        params += [f"--bind={config.build_sources}:/root/src{idmap_opt}"]
-
-    return params
+def script_env_params(state: MkosiState) -> list[str]:
+    return flatten(("--setenv", k, v) for k, v in state.environment.items())
```
This should be moved to `run.py`, so `__init__.py` gets even emptier.

DaanDeMeyer replied:
Moved to backend, I think it fits better there

behrmann commented on mkosi/__init__.py:
```diff
@@ -3799,13 +3536,26 @@ def ensure_networkd(config: MkosiConfig) -> bool:
+                                            text=True).stderr
+
+
 def run_shell(config: MkosiConfig) -> None:
```
Any reason not to move `run_shell` into `run.py` as well?

DaanDeMeyer replied:
Depends on `MkosiConfig` and we can't import `backend` into `run`. Doing this requires splitting things up more, so I'm not doing it for now

DaanDeMeyer replied:
PR is big already, let's leave moving the other run() functions in for another PR

behrmann commented on mkosi/__init__.py:
```diff
@@ -4334,9 +4079,71 @@ def needs_build(config: Union[argparse.Namespace, MkosiConfig]) -> bool:
+    return int(start), int(count)
+
+
+def become_root() -> None:
```
I think this would also be a good candidate for `run.py`.

behrmann reviewed (commented):
Looks really good! Some nits about placement of functions. This cuts so much stuff out of `__init__.py` so maybe the new stuff can be added elsewhere as well.

Gentoo still has permission errors.

DaanDeMeyer commented:
> Gentoo still has permission errors.

Yeah Gentoo is going to need a big rework. We currently use system directories all over the place. This should all be changed to use the workspace instead. I also think we should move back to using `emerge` as a command line tool instead of a python import so we can run it using bubblewrap with `run_with_apivfs()`.

I really don't have the time or expertise to do this Gentoo rework so I hope @257 can take care of that. In the meantime I'd suggest we disable Gentoo in CI so we can land this.

DaanDeMeyer commented:
One more change. Instead of mapping the current user to the same uid in the user namespace, we map it to the last uid/gid in the subuid/subgid ranges from /etc/subuid and /etc/subgid. This should be safer and pretty much ensure we'll never be using a uid/gid that's actually used in the image itself as well.

behrmann commented on mkosi/__init__.py:
```diff
@@ -1163,15 +983,18 @@ def calculate_bmap(state: MkosiState) -> None:
 
 
+def acl_allow_remove(root: Path, uid: int) -> None:
+    run(["setfacl", "-R", "-m", f"user:{uid}:rwx", root,
```
By default a recursive setfacl will follow symlink arguments, but not symlinks it encounters later. If you can't think of a good reason against it, I'd rather throw in a `--physical` here just to be on the safe side.

Also, we should probably be lenient here, if this fails and just emit a warning, that people will have to clean up their stuff with elevated privileges, since I'm sure somebody will try this on an FS that can't do ACLs or only has them as opt-in (i.e. zfs).

behrmann commented:
LGTM, cleared the CI warnings. This is a pretty big one, though, so somebody else should look at it, too.

behrmann commented on mkosi.md:
```diff
@@ -792,32 +782,31 @@ a machine ID.
-  found in the local directory it is automatically used for this
-  purpose (also see the "Files" section below). Specify an empty value
-  to disable automatic detection.
+  development image and executed inside a chroot environment. If this
```
I think we should be explicit here (and in the rest of the docs, where it says `chroot`), that we're using bubblewrap, just so that people know that the assurances are stronger than with a plain chroot.

DaanDeMeyer replied:
The thing is, bubblewrap on its own doesn't say much, since its a pretty generic tool

behrmann replied:
That's true, although this might be a tad too vague, how about
```suggestion
  development image and executed inside a namespaced chroot environment. If this
```

It's not really ideal, since it doesn't say anything about what namespaces are being used, but reading it, will hopefully give people the idea that it's more than a simple chroot.

behrmann commented on mkosi/run.py:
```diff
@@ -0,0 +1,273 @@
+    for line in path.read_text().splitlines():
+        name, start, count = line.split(":")
+
+        if name == pwd.getpwuid(os.getuid()).pw_name:
```
That's maybe an esoteric worry, but if the file contains UIDs that are not mapped on the host (where we run `pwd.getpwuid`) this will throw a `KeyError`.

behrmann commented on mkosi/run.py:
```diff
@@ -0,0 +1,273 @@
+
+        newuidmap = [
+            "newuidmap", pid,
+            0, subuid, subuidcount - 100,
```
What if `subuidcount` is smaller than 100? If we disallow that, this should be caught earlier.

DaanDeMeyer replied:
I added enforcing that the range is of size 65536

bluca commented:
> For apt, to accomplish this, we switch from using
DPkg::Chroot-Directory to setting the "--root" option for each
invocation of dpkg so that dpkg can access files outside of the
chroot.

Uhh - very few maintainer scripts actually support `$DPKG_ROOT` - are you sure this is working as intended and it's not running postinst/postrm scripts on the host filesystem?

DaanDeMeyer commented:
> > For apt, to accomplish this, we switch from using
> DPkg::Chroot-Directory to setting the "--root" option for each
> invocation of dpkg so that dpkg can access files outside of the
> chroot.
> 
> Uhh - very few maintainer scripts actually support `$DPKG_ROOT` - are you sure this is working as intended and it's not running postinst/postrm scripts on the host filesystem?

The option is documented as dpkg only chrooting for the install and for running the scripts. The entire thing is also running unprivileged, if any scripts were running in the host filesystem, I'd be running into permission errors (which I did when playing with --force-script-chrootless but not without it)

behrmann commented on mkosi/backend.py:
```diff
@@ -874,17 +543,12 @@ def sort_packages(packages: Iterable[str]) -> list[str]:
+
+
+def current_user_uid_gid() -> tuple[int, int]:
+    uid = int(os.getenv("SUDO_UID") or os.getenv("PKEXEC_UID") or str(os.getuid()))
```
```suggestion
    uid = int(os.getenv("SUDO_UID") or os.getenv("PKEXEC_UID") or os.getuid())
```
`int()` is idempotent for ints, so not need to roundtrip this via `str()`

behrmann commented on mkosi/run.py:
```diff
@@ -0,0 +1,313 @@
+    run(["mount", "--make-rslave", "/"])
+
+
+def foreground() -> None:
+    # If we're connected to a terminal, put the the child process in a new process group and make that
+    # the foreground process group so that only the child process receives SIGINT.
```
```suggestion
def foreground() -> None:
    """
    If we're connected to a terminal, put the the child process in a new process group and make that
    # the foreground process group so that only the child process receives SIGINT.
    """
```
Talking about the child process here is a bit confusing.

behrmann commented on mkosi/run.py:
```diff
@@ -0,0 +1,313 @@
+    return SUBRANGE - 100, SUBRANGE - 100
+
+
+def init_namespace() -> None:
```
```suggestion
def init_mount_namespace() -> None:
    """Set up a mount namespace and one-way recursive mount propagation from the host"""
```

behrmann commented on mkosi/run.py:
```diff
@@ -0,0 +1,313 @@
+    return int(start)
+
+
+def become_root() -> tuple[int, int]:
```
```suggestion
def become_root() -> tuple[int, int]:
    """Set up a new user namespace mapping using /etc/subuid and /etc/subgid.
    
    The current user will be mapped to root and 65436 will be mapped to the UID/GID of the invoking user. 
    The other IDs will bemapped through.
    
    The function returns the UID-GID pair of the invoking user in the namespace (65436, 65436).
    """
```
Shame docstrings can't be f-strings, Would be nice if the docs could be self-updating.

behrmann commented on mkosi/run.py:
```diff
@@ -0,0 +1,313 @@
+    else:
+        die(f"No mapping found for {user or uid} in {path}")
+
+    if int(count) != SUBRANGE:
```
That seems a bit strong. Maybe
```suggestion
    if int(count) < SUBRANGE:
```
so that it's at least this wide and we only handle the window of length `SUBRANGE`?

behrmann commented on mkosi/run.py:
```diff
@@ -0,0 +1,313 @@
+        signal.signal(signal.SIGTTOU, old)
+
+
+# Stores the exception from a subprocess along with its traceback. We have to do this explicitly because the
+# original traceback object cannot be pickled. When stringified, produces the subprocess stacktrace plus the
+# exception message.
+class RemoteException(Exception):
```
```suggestion
class RemoteException(Exception):
    """
    Stores the exception from a subprocess along with its traceback. We have to do this explicitly because the
    original traceback object cannot be pickled. When stringified, produces the subprocess stacktrace plus the
    exception message.
    """
```

behrmann commented on mkosi/run.py:
```diff
@@ -0,0 +1,313 @@
+        return f"Traceback (most recent call last):\n{''.join(self.tb.format()).strip()}\n{type(self.exception).__name__}: {self.exception}"
+
+
+# Attach to sys.excepthook to automically format exceptions with a RemoteException attached correctly.
+def excepthook(exctype: Type[BaseException], exc: BaseException, tb: Optional[TracebackType]) -> None:
```
```suggestion
def excepthook(exctype: Type[BaseException], exc: BaseException, tb: Optional[TracebackType]) -> None:
    """Attach to sys.excepthook to automically format exceptions with a RemoteException attached correctly."""
```

behrmann commented on mkosi/run.py:
```diff
@@ -0,0 +1,313 @@
+        sys.__excepthook__(exctype, exc, tb)
+
+
+def fork_and_wait(target: Callable[[], T]) -> T:
```
```suggestion
def fork_and_wait(target: Callable[[], T]) -> T:
    """Run target function in a child process and collect its backtrace if there is one."""
```

behrmann commented on mkosi/__init__.py:
```diff
@@ -4356,16 +4116,18 @@ def run_verb(raw: argparse.Namespace) -> None:
-            init_namespace()
-            build_stuff(config)
+            def target() -> None:
+                uid, gid = become_root() if os.getuid() != 0 else current_user_uid_gid()
```
```suggestion
                # Get the user UID/GID either on the host or in user ns running the build
                uid, gid = become_root() if os.getuid() != 0 else current_user_uid_gid()
```

behrmann reviewed (commented):
This is some really great work with some gnarly low-level Python.

The suggestions are mostly cosmetic stuff, to move comments to docstrings and add some more comments and docstrings to make this code easier to follow for the people who haven't written it. :)

behrmann commented on mkosi/run.py:
```diff
@@ -0,0 +1,326 @@
+    else:
+        die(f"No mapping found for {user or uid} in {path}")
+
+    if int(count) <= SUBRANGE:
```
Now if the range is exactly 2¹⁶ this will break.

bluca commented:
> > > For apt, to accomplish this, we switch from using
> > > DPkg::Chroot-Directory to setting the "--root" option for each
> > > invocation of dpkg so that dpkg can access files outside of the
> > > chroot.
> > 
> > 
> > Uhh - very few maintainer scripts actually support `$DPKG_ROOT` - are you sure this is working as intended and it's not running postinst/postrm scripts on the host filesystem?
> 
> The option is documented as dpkg only chrooting for the install and for running the scripts. The entire thing is also running unprivileged, if any scripts were running in the host filesystem, I'd be running into permission errors (which I did when playing with --force-script-chrootless but not without it)

The problem is that all scripts must be updated to the use $DPKG_ROOT env var, and most are not, because the scope for that is to allow cross-bootstrapping new architectures, so only a select few packages in the essential set have been updated.
So, I'd recommend to add a check that euid != 0 before doing that, because if it is ran as root it might do some serious damage to the host. Yes, this sucks.

DaanDeMeyer commented:
> The problem is that all scripts must be updated to the use $DPKG_ROOT env var, and most are not, because the scope for that is to allow cross-bootstrapping new architectures, so only a select few packages in the essential set have been updated.
So, I'd recommend to add a check that euid != 0 before doing that, because if it is ran as root it might do some serious damage to the host. Yes, this sucks.

But the thing is that with this option, dpkg still chroots() before invoking a maintainer script. So they don't need to support `DPKG_ROOT`. With the apt option `DPkg::Chroot-Directory`, apt does the chroot before invoking dpkg, so dpkg can't access outside the chroot at all. When using `--root`, dpkg will chroot itself, and only when needed.

bluca commented:
> > The problem is that all scripts must be updated to the use $DPKG_ROOT env var, and most are not, because the scope for that is to allow cross-bootstrapping new architectures, so only a select few packages in the essential set have been updated.
> > So, I'd recommend to add a check that euid != 0 before doing that, because if it is ran as root it might do some serious damage to the host. Yes, this sucks.
> 
> But the thing is that with this option, dpkg still chroots() before invoking a maintainer script. So they don't need to support `DPKG_ROOT`. With the apt option `DPkg::Chroot-Directory`, apt does the chroot before invoking dpkg, so dpkg can't access outside the chroot at all. When using `--root`, dpkg will chroot itself, and only when needed.

ok, I guess I'm getting confused as they are named in the same way, but if you double-checked that chroot is always used then it's fine

257 commented:
@DaanDeMeyer you fixed gentoo too!? it works for me here, i'm updating `mkosi.ebuild`. thanks.

DaanDeMeyer commented:
@257 I didn't fix Gentoo unfortunately. Can you verify you can build Gentoo images without root from main?

257 commented:
it builds here, i only had to install `bubblewrap`. are tests disabled on main? maybe we should test with CI and see what breaks and i will pick it up from there. but here, on a gentoo machine, build finished fine. didn't test for qemu though.

DaanDeMeyer commented:
Can you make a PR to enable gentoo again, that should surface some failures.

257 commented:
will do in a minute, qemu works here too btw.
